From 0022b237eed633268c383d80c6f65184a8a57ce4 Mon Sep 17 00:00:00 2001
From: Stuart Morgan <smorgan@mythtv.org>
Date: Tue, 8 Jul 2014 10:22:00 +0100
Subject: [PATCH 56/56] MythMusic: Read/Set an annonymous POPM tag in ID3
 tags. Let's us update the playcount/rating in MythTV
 and have that change reflected in Amarok and other
 applications that look for an anonymous POPM tag.

---
 mythtv/libs/libmythmetadata/metaioid3.cpp |   41 +++++++++++++++++++++++++----
 1 file changed, 36 insertions(+), 5 deletions(-)

diff --git a/mythtv/libs/libmythmetadata/metaioid3.cpp b/mythtv/libs/libmythmetadata/metaioid3.cpp
index 02388a3..3dae5cd 100644
--- a/mythtv/libs/libmythmetadata/metaioid3.cpp
+++ b/mythtv/libs/libmythmetadata/metaioid3.cpp
@@ -286,9 +286,16 @@ MusicMetadata *MetaIOID3::read(const QString &filename)
         metadata->setCompilationArtist(compilation_artist);
     }
 
-    // MythTV rating and playcount, stored in POPM frame
-    PopularimeterFrame *popm = findPOPM(tag, email);
+    // Rating and playcount, stored in POPM frame
+    PopularimeterFrame *popm = findPOPM(tag, ""); // Global (all apps) tag
 
+    // If no 'global' tag exists, look for the MythTV specific one
+    if (!popm)
+    {
+        popm = findPOPM(tag, email);
+    }
+
+    // Fallback to using any POPM tag we can find
     if (!popm)
     {
         if (!tag->frameListMap()["POPM"].isEmpty())
@@ -833,6 +840,7 @@ bool MetaIOID3::writePlayCount(TagLib::ID3v2::Tag *tag, int playcount)
     if (!tag)
         return false;
 
+    // MythTV Specific playcount Tag
     PopularimeterFrame *popm = findPOPM(tag, email);
 
     if (!popm)
@@ -842,7 +850,18 @@ bool MetaIOID3::writePlayCount(TagLib::ID3v2::Tag *tag, int playcount)
         popm->setEmail(email);
     }
 
-    popm->setCounter(playcount);
+    int prevCount = popm->counter();
+    int countDiff = playcount - prevCount;
+    // Allow for situations where the user has rolled back to an old DB backup
+    if (countDiff > 0)
+        popm->setCounter(playcount);
+
+    // Global playcount Tag - Updated by all apps/hardware that support it
+    if (countDiff > 0)
+    {
+        PopularimeterFrame *gpopm = findPOPM(tag, "");
+        gpopm->setCounter(gpopm->counter() + countDiff);
+    }
 
     return true;
 }
@@ -876,6 +895,10 @@ bool MetaIOID3::writeRating(TagLib::ID3v2::Tag *tag, int rating)
     if (!tag)
         return false;
 
+    int popmrating = static_cast<int>(((static_cast<float>(rating) / 10.0)
+                                                               * 255.0) + 0.5);
+
+    // MythTV Specific Rating Tag
     PopularimeterFrame *popm = findPOPM(tag, email);
 
     if (!popm)
@@ -884,10 +907,18 @@ bool MetaIOID3::writeRating(TagLib::ID3v2::Tag *tag, int rating)
         tag->addFrame(popm);
         popm->setEmail(email);
     }
-    int popmrating = static_cast<int>(((static_cast<float>(rating) / 10.0)
-                                                               * 255.0) + 0.5);
     popm->setRating(popmrating);
 
+    // Global Rating Tag
+    PopularimeterFrame *gpopm = findPOPM(tag, "");
+    if (!gpopm)
+    {
+        gpopm = new PopularimeterFrame();
+        tag->addFrame(gpopm);
+        gpopm->setEmail("");
+    }
+    gpopm->setRating(popmrating);
+
     return true;
 }
 
-- 
1.7.10.2

